File: backend/app/schemas/debate.py
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


VerdictOutcome = Literal["APPROVE", "REJECT", "ESCALATE TO HUMAN"]
AgentRole = Literal["hermes", "nemesis"]
Stance = Literal["approve", "reject", "escalate"]
SseEventType = Literal[
    "status",
    "specter_brief",
    "agent_message",
    "round_state",
    "verdict",
    "error",
    "done",
]


class DebateRequest(BaseModel):
    raw_transaction: str = Field(min_length=1)
    max_rounds: int = Field(default=6, ge=1, le=6)
    escalation_email: Optional[str] = None


class ParsedTransaction(BaseModel):
    raw: str
    date: Optional[str] = None
    vendor: str
    amount: Optional[float] = None
    currency: Optional[str] = None
    reference: Optional[str] = None
    notes: Optional[str] = None


class SpecterBrief(BaseModel):
    status: Literal["available", "unavailable"]
    summary: str
    vendor_found: bool = False
    vendor_name: Optional[str] = None
    domain: Optional[str] = None
    founded_year: Optional[int] = None
    legitimacy_score: Optional[float] = None
    red_flags: List[str] = Field(default_factory=list)
    raw: Optional[Dict[str, Any]] = None


class AgentStructuredMessage(BaseModel):
    role: AgentRole
    round: int = Field(ge=1)
    stance: Stance
    claim: str
    evidence: str
    counterpoint: str
    risk_tags: List[str] = Field(default_factory=list)
    confidence: float = Field(ge=0.0, le=1.0)
    text: str
    structured_available: bool = True

    @field_validator("confidence")
    @classmethod
    def _round_confidence(cls, v: float) -> float:
        return round(float(v), 4)


class VerdictDecision(BaseModel):
    outcome: VerdictOutcome
    reason: str
    escalation_packet: Optional[Dict[str, Any]] = None


class SseEnvelope(BaseModel):
    event_id: str
    trace_id: str
    event_type: SseEventType
    timestamp: str
    payload: Dict[str, Any]

File: frontend/src/types/debate.ts
export type AgentRole = "hermes" | "nemesis";
export type Stance = "approve" | "reject" | "escalate";
export type VerdictOutcome = "APPROVE" | "REJECT" | "ESCALATE TO HUMAN";

export type SseEventType =
  | "status"
  | "specter_brief"
  | "agent_message"
  | "round_state"
  | "verdict"
  | "error"
  | "done";

export interface ParsedTransaction {
  raw: string;
  date?: string | null;
  vendor: string;
  amount?: number | null;
  currency?: string | null;
  reference?: string | null;
  notes?: string | null;
}

export interface SpecterBrief {
  status: "available" | "unavailable";
  summary: string;
  vendor_found: boolean;
  vendor_name?: string | null;
  domain?: string | null;
  founded_year?: number | null;
  legitimacy_score?: number | null;
  red_flags: string[];
}

export interface AgentStructuredMessage {
  role: AgentRole;
  round: number;
  stance: Stance;
  claim: string;
  evidence: string;
  counterpoint: string;
  risk_tags: string[];
  confidence: number;
  text: string;
  structured_available: boolean;
}

export interface VerdictDecision {
  outcome: VerdictOutcome;
  reason: string;
  escalation_packet?: Record<string, unknown> | null;
}

export interface SseEnvelope<T = Record<string, unknown>> {
  event_id: string;
  trace_id: string;
  event_type: SseEventType;
  timestamp: string;
  payload: T;
}

export interface DebateRequest {
  raw_transaction: string;
  max_rounds?: number;
}
